r-data-manipulation.Rmd

```diff
@@ -0,0 +1,1111 @@
+
+This is an incredibly common way of *displaying* data, but when *storing* tabular data in a file, we need a way to communicate when records and values begin and end.  A very popular format for doing this is CSV.  CSV, is short for **c**omma **s**eparated **v**alues, and like the name suggests, a comma, `,`, is used to separate the values for each column, while each record goes on a new line.  The file is plain text, but we use the extension `.csv` to indicate that is follows the CSV format conventions.
+
+Here's how the table above would look inside the CSV file `site.csv`:
+```{r, echo = FALSE, comment = NA}
+writeLines(readLines(example_csv))
+```
+
+In this case the first line has the column names, this is common (and recommended!), but not universal. 
+
+Why is CSV so popular? 
+
+* It's human readable.  CSV isn't a special file type, it is just a plain text file that follows some conventions.  This means you don't need any special software to look at the contents -- you can open in up in anything that can examine text and take a look inside. 
+* It's computer readable.  Because CSV files all have the same structure it's easy to write computer programs to read them.  This means in almost any program designed to work with data, and all the common programming languages, you'll find functions that will import CSV files.  This also means it's easy to create CSV files -- you can export them from Excel, write them from R, or even write one from scratch in a text editor.
+
+If you want to look inside a CSV file in RStudio you can navigate to its location in the "Files" pane and click on its name.  Selecting "View File", will open it in the Source pane. 
+
+However, if you want to work with CSV data in R, it isn't enough to look inside the file.  You need to read the contents of the file and store it in R's memory.  This process is known as data import.
+
+### Importing CSV data into R
+
+To work with data in R you need to have it in R's memory.  The `read_csv()` function in the readr package will import a CSV file, and represent it as a tibble, if you give it the location of the CSV file.  For example, to read the `site.csv` data and store it in an object called `sites`: 
+```{r}
+library(tidyverse)
+sites <- read_csv(here("data", "site.csv"))
+sites
+```
+
+TODO: talk about files paths, point reader to the place where file paths are talked about, or assume file is in their working directory.
+
+Notice that `read_csv()` gave us a message about what it did: it parsed our data file and found three columns `site_id`, `latitude` and `longitude`.  It also mentions what kind of data it assumed was in each column.  TODO: point to further discussion of data types.
+
+The object `sites` is now R's representation of the data from the `site.csv` file.    
+
+### Exercise: Import `visited.csv`
+
+Use `read_csv()` to read `visited.csv` into R. How does R indicate a cell with a missing value?
```

Comment:
I assume they'll have access to this dataset? How do we get them to download it, etc?

Reply:
Good question.  Do we:

* link to it, and require a student to download to an appropriate location, 
* give code to download it to an appropriate location, 
* have the raw data as part of a package that a student has installed and give code to get its path 
* rely on the "instructor" to provide it (i.e. in a rstudio cloud project)?

Other ideas? Thoughts?

Reply:
I think we package it...